### PR TITLE
feat: add comprehend pii

### PR DESCRIPTION
**Pull request by MichaelWalker-git**
### Description
- Add Comprehend PII Construct

### Ticket
https://github.com/aws-samples/amazon-textract-idp-cdk-constructs/issues/50

### CODE REVIEW

File: lambda/comprehend_pii_sync/app/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+amazon-textract-caller
```
Comment (schadem) [inline lambda/comprehend_pii_sync/app/requirements.txt:1]: amazon-textract-caller is not used

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+from botocore.exceptions import ClientError
+from typing import List, Tuple
+
+logger = logging.getLogger(__name__)
+version = "0.0.1"
+
+s3 = boto3.client('s3')
+step_functions_client = boto3.client(service_name='stepfunctions')
+sqs = boto3.client('sqs')
```
Comment (schadem) [inline lambda/comprehend_pii_sync/app/sync_main.py:19]: is not used

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                token: {token} \n \
+                execution_id: {execution_id}")
+
+    try:
+        text = get_file_from_s3(s3_path=s3_input_text).decode('utf-8')[0:4900]
+
+        start_time = round(time.time() * 1000)
+        response = comprehend.detect_pii_entities(
+            Text=text,
+            EndpointArn=comprehend_classifier_arn,
```
Comment (schadem) [inline lambda/comprehend_pii_sync/app/sync_main.py:137]: not supported by detect_pii_entities

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            LanguageCode=language_code,
+            DocumentReaderConfig={
+                'DocumentReadAction': 'TEXTRACT_DETECT_DOCUMENT_TEXT',
+                'DocumentReadMode': 'SERVICE_DEFAULT',
+                'FeatureTypes': [
+                    'TABLES', 'FORMS',
+                ]
+            }
```
Comment (schadem) [inline lambda/comprehend_pii_sync/app/sync_main.py:145]: not supported by detect_pii_entities

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        )
+        logger.debug(f"comprehend result: {response}")
+
+        classification_result = "NONE"
+        for c in response['Classes']:
+            if c['Score'] > 0.50:
+                classification_result = c['Name']
+                break
```
Comment (schadem) [inline lambda/comprehend_pii_sync/app/sync_main.py:153]: is not a classifier

**Review by schadem — CHANGES_REQUESTED** [review]
This does not seem to work

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        logger.error(f"TaskDoesNotExist for message: {message} ")
+    except step_functions_client.exceptions.TaskTimedOut:
+        logger.error(f"TaskTimedOut for message: {message} ")
+
+
+def send_success_to_step_function(response, token, event, start_time):
+    logger.debug(f"comprehend result: {response}")
+
+    classification_result = "NONE"
+    for c in response['Entities']:
```
Comment (MichaelWalker-git) [inline lambda/comprehend_pii_sync/app/sync_main.py:65]: Input: Textract JSON
Output - PII:
- Enhanced Textract json (list of blocks: words and lines)
- Indicate which words are PII
- Remapping using BeginOffset 
-

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    logger.info(f"s3_path: {s3_path} \n \
+                    s3_input_text: {s3_input_text} \n\
+                token: {token} \n \
+                execution_id: {execution_id}")
+
+    try:
+        text = get_file_from_s3(s3_path=s3_input_text).decode('utf-8')[0:4900]
+
+        start_time = round(time.time() * 1000)
+        response = comprehend.detect_pii_entities(
```
[the remark this answers is not in the document]
Comment (MichaelWalker-git) [reply lambda/comprehend_pii_sync/app/sync_main.py:126]: Mark the block as PII. Custom element

File: lambda/comprehend_pii_sync/app/sync_main.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            Text=text,
+            LanguageCode=language_code,
+        )
+        send_success_to_step_function(response, token, event, start_time)
```
Comment (MichaelWalker-git) [inline lambda/comprehend_pii_sync/app/sync_main.py:130]: All the blocks of textract + information about the offset.

Each words would have an offset.

File: lambda/comprehend_pii_sync/app/sync_main.py (hunk 6 above)
Comment (MichaelWalker-git) [reply lambda/comprehend_pii_sync/app/sync_main.py:65]: https://github.com/aws-samples/amazon-textract-idp-cdk-constructs/blob/3f40a57261d75c7a37d9a375472c816327fcad93/lambda/generatecsv/app/main.py#L96

File: lambda/comprehend_pii_sync/app/sync_main.py (hunk 8 above)
Comment (MichaelWalker-git) [reply lambda/comprehend_pii_sync/app/sync_main.py:130]: https://github.com/aws-samples/amazon-textract-idp-cdk-constructs/blob/3f40a57261d75c7a37d9a375472c816327fcad93/lambda/generatecsv/app/main.py#L96

**Pull request closed without merging**